content/post/smaller-the-parameter-smaller-the-bugs.md

```diff
@@ -0,0 +1,243 @@
+
+#### Before
+```rb
+def summary(first_name, last_name)
+  "#{title} - #{last_name}, #{first_name} (#{publish_year})"
+end
+ 
+book = Book.new("Ruby is the best", "2020")
+book.summary("Bater", "Chen")
+=> "Ruby is the best - Chen, Bater (2020)"
+```
+#### After
+```rb
+def summary(author)
+  "#{title} - #{author.full_name} (#{publish_year})"
+end
+ 
+book = Book.new("Ruby is the best", "2020")
+book.summary(Author.new("Bater", "Chen"))
+=> "Ruby is the best - Chen, Bater (2020)"
+```
+The output of `summary` is the same, but the parameter reduces from two strings to one object. Well, you may doubt that it doesn't really matter, but imagine that you can follow the same pattern to reduce the parameter from 10 data into 1 object. That would be a huge impact.
+
+## Law of Demeter
+You may also notice that I relocate the logic of full name from book class to author class because a full name can perform independently without interaction with a book object. An **Object should know less about each other**, it is known as the [Law of Demeter][2] or "Least Knowledge Principle".
+
+In this case, the book class shouldn't know the logic of `full_name`. The business logic of `full_name` should only be defined in the author class, not spread to somewhere else. From now on, the book class has no dependence on first name and last name directly. In fact, any object can be the parameter of the summary method once it has the `full_name` method. For example, if we have a user object that also can return `full_name`, it also works fine with the `summary` method.
+
+This practice is also known as [Duck Typing][1]. If it walks like a duck and it quacks like a duck, then it must be a duck. In duck typing, an object's suitability is determined by the presence of certain methods and properties, rather than the type of the object itself. In the Java case, you can use the interface.
+
+Now a new requirement is coming, we need to add a new method called `cover`, and it also has author first name and last name inside. We can follow the pattern we just made. Let's compare before and after the refactoring of the book class.
+#### Old pattern without refactoring
+```rb
+def summary(first_name, last_name)
+  "#{title} - #{last_name}, #{first_name} (#{publish_year})"
+end
+ 
+def cover(first_name, last_name)
+  "#{title}\n\n\n#{last_name}, #{first_name}"
+end
+```
+#### After refactor
+```rb
+def summary(author)
+  "#{title} - #{author.full_name} (#{publish_year})"
+end
+ 
+def cover(author)
+  "#{title}\n\n\n#{author.full_name}"
+end
+```
+The cover method generates a string that has the book title and author's full name inside. After the refactoring, we can reuse the author's full name method again. Once we build a good pattern, the new code can follow it easily and keep the project clean. But if you didn't pay attention, the code would get rust rapidly and becomes the developer's nightmare. 
```

Comment:
```suggestion
The cover method generates a string that has the book title and author's full name inside. After the refactoring, we can reuse the author's full name method again. Once we build up a good pattern, the new code can follow it easier and keep our project clean. But if you didn't pay attention, the code would quickly become harder to read, develop  technical debt and become a developer's worst nightmare. 
```
Wasn't sure how to correct "code get rusty rapidly". You might need to reword this